README.md
# ExtJs Package Wrapper for plyr

[![perry-johnson](https://img.shields.io/badge/perry%20johnson-pja-blue.svg)](https://www.perryjohnson.com)
[![app-type](https://img.shields.io/badge/category-linux%20web%20env-blue.svg)](https://www.perryjohnson.com)
[![app-lang](https://img.shields.io/badge/language-na%20c%23-blue.svg)](https://www.perryjohnson.com)
[![app-publisher](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-app--publisher-e10000.svg)](https://github.com/perryjohnsoninc/app-publisher)
[![authors](https://img.shields.io/badge/authors-scott%20meesseman-6F02B5.svg?logo=visual%20studio%20code)](https://github.com/perryjohnsoninc)

## Description

> This package provides an ExtJS package wrapper for the [plyr html5 media player](https://github.com/sampotts/plyr) by [Sam Potts](https://github.com/sampotts), available on [npmjs.org](https://www.npmjs.com/package/plyr).  The plyr package is used as a dependency and this package will include its distribution files into ExtJs client builds.

## Install

To install this package, run the following command:

    npm install @perryjohnson/extjs-pkg-plyr

## Usage

To include the package in an ExtJS application build, be sure to add the package name to the list of required packages in the app.json file:

    "requires": [
         "plyr",
        ...
    ]

For an open tooling build, also add the node_modules path to the workspace.json packages path array:

     "packages": {
        "dir": "...${package.dir}/node_modules/@perryjohnson/extjs-pkg-plyr"
    }

Simply include the control into any class file:

    require: [ 'Ext.ux.Plyr' ],
    items: [
    {
        xtype: 'plyr',
        audioCtlListTags: 'download',
        currentTime: 0,
        url: 'https://www.mydomain.com/audio/blank.mp4',
        plyrLoaded: function()         // Optional callback
        {
            Utils.log('Loaded!!!');
        },
        plyrLog: function(msg, level)  // Optional callback
        {
            Utils.log(msg, level);
        }
    }]


CHANGELOG.md
# EXTJS-PKG-PLYR CHANGE LOG

## Version 1.2.2 (July 18th, 2019)

### Build System

- **ap:** set history file to empty for git test
- **ap:** set interactive flag back to N
- **ap:** update publishrc to 1.7 config
- **npm:** update bugs page and homepage to latest
- **npm:** update npmrc

### Documentation

- **readme:** update badge info and usage info
- **readme:** update badges

### Code Refactoring

- audio player should count up instead of down.

	[closes #161]

- use lf for package.json/pakage-lock.json for npm compatibility

### Code Styling

- cleanup. remove semantic-release dependencies

## Version 1.2.1 (May 28th, 2019)

### Build System

- **npm:** fix app-publisher script arguments
- **npm:** fix invalid attribute publisher->author [skip ci]
- **npm:** package.json cleanup [skip ci]
- **npm:** switch to perryjohnson scope from testing scope
- modify script to replace speed rates in plyr source

### Chores

- 1.0.11 post release check in
- 1.2.0 release update
- add node_modules folder to ignore list
- prep file check in for first app-publisher release
- set scope name

### Features

- add additional speed playback rates to menuSpeed rates are now 0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.25, 1.5, 1.75, 2
- initial check in tested npm package project

### Bug Fixes

- include plyr release files in package

### Code Styling

- ignore unused files

### Tests

- scoped package

### Other Notes

- scoped package

# [1.2.0](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.1.0...v1.2.0) (2019-05-04)


### Features

* add additional speed steps 0.6, 0.7, 0.8, 0.9 ([0ecc9b6](https://github.com/spmeesseman/extjs-pkg-plyr/commit/0ecc9b6))

# [1.1.0](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.15...v1.1.0) (2019-05-03)


### Features

* emit onLoaded and onProgress events for ui to consume ([ada5c13](https://github.com/spmeesseman/extjs-pkg-plyr/commit/ada5c13))

## [1.0.15](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.14...v1.0.15) (2019-05-03)


### Bug Fixes

* copy plyr dist from new local package dir on sencha app build ([1d18cf9](https://github.com/spmeesseman/extjs-pkg-plyr/commit/1d18cf9))

## [1.0.14](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.13...v1.0.14) (2019-05-03)


### Bug Fixes

* include plyr files in package, not as separate dependency ([e1b9f0a](https://github.com/spmeesseman/extjs-pkg-plyr/commit/e1b9f0a))


### Build System

* **npm:** back to non-scoped conventional-changelog-spm ([e2372cc](https://github.com/spmeesseman/extjs-pkg-plyr/commit/e2372cc))

## [1.0.13](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.12...v1.0.13) (2019-05-02)


### Bug Fixes

* include plyr files in package, not as separate dependency ([0ac7e3f](https://github.com/spmeesseman/extjs-pkg-plyr/commit/0ac7e3f))

## [1.0.12](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.11...v1.0.12) (2019-05-01)


### Code Refactoring

* make package scoped ([6180d14](https://github.com/spmeesseman/extjs-pkg-plyr/commit/6180d14))

## [1.0.11](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.10...v1.0.11) (2019-04-30)


### Code Refactoring

* remove build.xml from release package ([a1684e5](https://github.com/spmeesseman/extjs-pkg-plyr/commit/a1684e5))


### Documentation

* **readme:** remove testing info ([81658a4](https://github.com/spmeesseman/extjs-pkg-plyr/commit/81658a4))

## [1.0.10](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.9...v1.0.10) (2019-04-30)


### Bug Fixes

* stop running rwd/ffwd timer on any new command ([79f4fc7](https://github.com/spmeesseman/extjs-pkg-plyr/commit/79f4fc7))


### Documentation

* **readme:** update builds list [skip ci] ([406aff8](https://github.com/spmeesseman/extjs-pkg-plyr/commit/406aff8))

## [1.0.9](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.8...v1.0.9) (2019-04-24)


### Bug Fixes

* undefined property Plyr when loading ([13bb3ec](https://github.com/spmeesseman/extjs-pkg-plyr/commit/13bb3ec))

## [1.0.8](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.7...v1.0.8) (2019-04-23)


### Bug Fixes

* first set of issues in first plyr implementation are resolved, first working built out api at extjs component level ([d834b3c](https://github.com/spmeesseman/extjs-pkg-plyr/commit/d834b3c))


### Build System

* **npm:** update security vulnerablities.  manual changes needed for tar vulnerability. ([c1f2e8d](https://github.com/spmeesseman/extjs-pkg-plyr/commit/c1f2e8d))
* **semantic-release:** add featmin keyword to commit-analyzer ([0d547c8](https://github.com/spmeesseman/extjs-pkg-plyr/commit/0d547c8))


### Code Refactoring

* change namespace/component name from Ext.plyr.Plyr to Ext.ux.Plyr ([774cf66](https://github.com/spmeesseman/extjs-pkg-plyr/commit/774cf66))


### Documentation

* **readme:** update build by me section ([f84abe3](https://github.com/spmeesseman/extjs-pkg-plyr/commit/f84abe3))

## [1.0.7](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.6...v1.0.7) (2019-04-19)


### Documentation

* **readme:** add Greenkeeper badge ([e661d77](https://github.com/spmeesseman/extjs-pkg-plyr/commit/e661d77))

## [1.0.6](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.5...v1.0.6) (2019-04-10)


### Bug Fixes

* cant load more than one player instance ([1656edf](https://github.com/spmeesseman/extjs-pkg-plyr/commit/1656edf))


### Documentation

* **readme:** update credits info of ply developer and project [skip ci] ([aadbf89](https://github.com/spmeesseman/extjs-pkg-plyr/commit/aadbf89))

## [1.0.5](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.4...v1.0.5) (2019-04-09)


### Bug Fixes

* expose dom playerid as a component property, null check log and onload callbacks ([eef183e](https://github.com/spmeesseman/extjs-pkg-plyr/commit/eef183e))

## [1.0.4](https://github.com/spmeesseman/extjs-pkg-plyr/compare/v1.0.3...v1.0.4) (2019-04-09)


### Bug Fixes


src/Plyr.js
Ext.define('Ext.ux.Plyr', 
{
    extend: 'Ext.Component',
    xtype: 'plyr',
	
	player: null,
	playerId: null,
    plyrInitialized: false,
    intializationInProgress: 0,
    reference: 'player',
    autoHeight: true,
    border:false,
    style: 
    {
    	padding: '0 0 0 0',
    	border: '0'
    },

	plyrOnLoaded: Ext.emptyFn,
	plyrOnProgress: Ext.emptyFn,
	plyrLog: Ext.emptyFn,

    statics:
    {
    	imgPath: 'resources/images',
		scriptLoaded: 0,
		playerIdCounter: 0,
    	captureActivity: 
    	{
    		fn: Ext.emptyFn,
    		scope: null
    	}
    },

    config:
    {
        url: '',
        idRoot: 0,
        audioCtlListTags: '',
        currentTime: 0,
        playbackSpeed: 1
    },

    publishes:
    {
        url: true,
        idRoot: true,
        audioCtlListTags: true,
        currentTime: true
    },

    getIdRoot: function()
    {
		var me = this;
		if (!me.idRoot) {
			me.idRoot = ++Ext.ux.Plyr.playerIdCounter;
			me.playerId = 'player_' + me.idRoot.toString();
		}
        return me.idRoot;
	},
	
	bind:
    {
        html: !Ext.isIE ? 
        //
        // Non-IE
        //
        '<audio id="player_{player.idRoot}" src="{player.url}" controls controlsList="{player.audioCtlListTags}" ' +
        'currenttime="{player.currentTime}" style="width:100%">This browser does not support HTML 5.</audio>' : 
        //
        // IE does not support HTML5 Audio Player
        //
        '<object id="player_{player.idRoot}" classid="clsid:02BF25D5-8C17-4B23-BC80-D3488ABDDC6B" '+
            'codebase="http://www.apple.com/qtactivex/qtplugin.cab" width="100%" height="50">' +
            '<param name="src" value="{player.url}">' +
            '<param name="autoplay" value="false">' +
            '<embed type="audio/x-wav" src="{player.url}" autoplay="false" autostart="false" width="100%" height="50">' +
        '</object>'
	},
	
	listeners:
	{
		destroy: function(cmp, eopts)
		{
			var me = this;
			if (me.player) {
				me.player.destroy();
			}
		}
	},

	//
	// List of available plyr API methods:
	//
	//     play()¹	           -	Start playback.
    //     pause()	           -	Pause playback.
    //     stop()	           -	Stop playback and reset to start.
    //     restart()	       -	Restart playback.
    //     fullscreen.enter()	-	Enter fullscreen. If fullscreen is not supported, a fallback "full window/viewport" is used instead.
    //     fullscreen.exit()	-	Exit fullscreen.
    //     fullscreen.toggle()	-	Toggle fullscreen.
    //     airplay()	        -	Trigger the airplay dialog on supported devices.
    //     destroy()	        -	Destroy the instance and garbage collect any elements.
    //     togglePlay(toggle)	    Boolean	Toggle playback, if no parameters are passed, it will toggle based on current status.
    //     rewind(seekTime)	        Number  Rewind playback by the specified seek time. If no parameter is passed, the default seek time will be used.
    //     forward(seekTime)	    Number	Fast forward by the specified seek time. If no parameter is passed, the default seek time will be used.
    //     increaseVolume(step)	    Number	Increase volume by the specified step. If no parameter is passed, the default step will be used.
    //     decreaseVolume(step)	    Number	Increase volume by the specified step. If no parameter is passed, the default step will be used.
    //     toggleCaptions(toggle)   Boolean	Toggle captions display. If no parameter is passed, it will toggle based on current status.
    //     toggleControls(toggle)   Boolean	Toggle the controls (video only). Takes optional truthy value to force it on/off.
    //     on(event, function)	    String, Function	Add an event listener for the specified event.
    //     once(event, function)    String, Function	Add an event listener for the specified event once.
    //     off(event, function)	    String, Function	Remove an event listener for the specified event.
    //     supports(type)	        String	Check support for a mime type.
	//


	taskRunner: null,
	taskRunnerTask: null,

	ffwd: function(seconds)
	{
		var me = this;
		
		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Fast Forward", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		if (!seconds) {
			seconds = 2;
		}

		me.stopTaskRunner();

		me.player.forward(seconds);
	},


	ffwdStart: function(seconds, stepms)
	{
		var me = this;

		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Fast Forward Start", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		me.stopTaskRunner();

		me.player.forward(seconds);

		//
		// TODO - this should actually use a webworker in the case the browser window
		// loses focus or is not the active window
		//
		if (!me.taskRunner) {
			me.taskRunner = new Ext.util.TaskRunner();	
		}

		me.taskRunnerTask = me.taskRunner.start({
			run: function()
			{
				me.player.forward(seconds);
			},
			interval: stepms ? stepms : 500
		});
	},


	ffwdRwdStop: function()
	{
		var me = this;

		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Fast Forward / Rewind Stop", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		me.stopTaskRunner();
	},


	pause: function()
	{
		var me = this;
		
		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Resume", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		me.stopTaskRunner();

		me.player.pause();
	},


	play: function()
	{
		var me = this;
		
		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Resume", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		me.stopTaskRunner();

		me.player.play();
	},


	togglePlay: function()
	{
		var me = this;
		
		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Play/Pause", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		me.stopTaskRunner();

		me.player.togglePlay();
	},


	rwd: function(seconds)
	{
		var me = this;
		
		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Rewind", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		if (!seconds) {
			seconds = 2;
		}

		me.stopTaskRunner();

		me.player.rewind(seconds);
	},


	rwdStart: function(seconds, stepms)
	{
		var me = this;

		if (me.plyrLog) {
			me.plyrLog("PLYR: Command received - Fast Forward", 1);
		}
		if (!me.player) {
			if (me.plyrLog) {
				me.plyrLog("PLYR: No active player", 1);
			}
			return;
		}

		me.stopTaskRunner();

		me.player.rewind(seconds);

		///
		// TODO - this should actually use a webworker in the case the browser window
		// loses focus or is not the active window
		//
		if (!me.taskRunner) {
			me.taskRunner = new Ext.util.TaskRunner();	
		}

		me.taskRunnerTask = me.taskRunner.start({
			run: function()
			{
				me.player.rewind(seconds);
			},
			interval: stepms ? stepms : 500
		});
	},


	stopTaskRunner: function()
	{
		var me = this;
		if (me.taskRunner && me.taskRunnerTask)
		{
			me.taskRunner.stop(me.taskRunnerTask, true);
			me.taskRunnerTask = null;
		}
	},


	afterRender: function() 
    {
    	var me = this;
    	me.callParent(arguments);

		if (me.plyrLog) {
			me.plyrLog("", 1);
		}

		if (Ext.isIE) {
			if (me.plyrLog) {
				me.plyrLog("HTML5 audio is not supported in IE.  Fall back to apple/qtplugin", 1);
			}
			return;
		}

		if (me.plyrLog) {
			me.plyrLog("Loading Plyr HTML5 Media", 1);
		}

    	//
    	// Load script
    	//
    	if (!Ext.ux.Plyr.sriptLoaded)
    	{
    		new Ext.util.DelayedTask(function()
    		{
    			Ext.Loader.loadScript(
    			{
    				url: Ext.manifest.resources.base + '/plyr/plyr.js',
    				//charset: 'UTF-8',
    				onLoad: me.onLoadSuccess,
    				onError: me.onLoadError,
    				scope: me
    			});

    		}, me).delay(5);
    	}
    	else
    	{
    		new Ext.util.DelayedTask(function()
    		{
				me.onLoadSuccess();
			}, me).delay(25);
    	}
    },


    //
    // Callback function for loading the TinyMCEscript
    //
    onLoadError: function() 
    {
		var me = this;
		if (me.plyrLog) {
			if (!Ext.ux.Plyr.sriptLoaded)
				me.plyrLog("   Error loading JS", 1);
			else if (Ext.ux.Plyr.sriptLoaded === 1)
				me.plyrLog("   Error loading CSS", 1);
		}
    },


    //
    // Calback function for successfully loading TinyMCE script
    //
    onLoadSuccess: function() 
    {
    	var me = this;
		if (!Ext.ux.Plyr.sriptLoaded) 
		{
			if (me.plyrLog) {
				me.plyrLog("   JS loaded", 1);
			}
			//
			// Load the CSS
			//
			Ext.ux.Plyr.sriptLoaded = 1;
			new Ext.util.DelayedTask(function()
    		{
    			Ext.Loader.loadScript(
    			{
    				url: Ext.manifest.resources.base + '/plyr/plyr.css',
    				//charset: 'UTF-8',
    				onLoad: me.onLoadSuccess,
    				onError: me.onLoadError,
    				scope: me
    			});
    		}, me).delay(5);
    	}
    	else if (Ext.ux.Plyr.sriptLoaded === 1)
    	{
			//
			// JS and CSS both loaded, initialize the plyr component
			//
			if (me.plyrLog) {
				me.plyrLog("   CSS loaded", 1);
			}
			Ext.ux.Plyr.sriptLoaded = true;
			me.onLoadSuccess(); // re-call with new scriptLoaded param set
    	}
		else if (Ext.ux.Plyr.sriptLoaded === true)
		{
			var opts = {
				enabled: true,           // Whether to completely disable Plyr
				debug: false,            // Display debugging information in the console
				autoplay: false,         // Autoplay the media on load. This is generally advised against
				autopause: true,         // Only allow one player playing at once
				muted: false,            // Whether to start playback muted
				hideControls: true,      // Hide video controls automatically after 2s of no mouse
				clickToPlay: true,       // Click (or tap) of the video container will toggle play/pause
				disableContextMenu: true,// Disable right click menu on video
				resetOnEnd: false,       // Reset the playback to the start once playback is complete
				seekTime: 10,            // The time, in seconds, to seek when a user hits fast forward or rewind
				volume: 1,               // A number, between 0 and 1, representing the initial volume of the player
				duration: null,          // Specify a custom duration for media.
				displayDuration: true,   // Displays the duration of the media on the "metadataloaded" event (on startup) in the current time display. This will only work if the preload attribute is not set to none (or is not set at all) and you choose not to display the duration (see controls option).
				invertTime: true,        // Display the current time as a countdown rather than an incremental counter
				//
				// If a function is passed, it is assumed your method will return either an element or HTML string 
				// for the controls. Three arguments will be passed to your function; id (the unique id for the 
				// player), seektime (the seektime step in seconds), and title (the media title). See controls.md 
				//for more info on how the html needs to be structured.
				// See https://github.com/spmeesseman/plyr/blob/master/controls.md for complete list of controls
				//
				controls: ['play-large', 'play', 'progress', 'current-time', 'mute', 'volume', 'captions', 'settings', 'pip', 'airplay', 'fullscreen'],
				// If using the default controls are used then you can specify which settings to show in the menu
				settings: ['captions', 'quality', 'speed', 'loop'],
				keyboard: { focused: true, global: false }, // Enable keyboard shortcuts for focused players only or globally
				tooltips: { controls: false, seek: true },  // Display control labels as tooltips on :hover & :focus (by default, the labels are screen reader only). seek: Display a seek tooltip to indicate on click where the media would seek to.
				storage:  { enabled: true, key: 'plyr' },   // enabled: Allow use of local storage to store user settings. key: The key name to use.
				captions: { active: false, language: 'auto', update: false },
				speed:    { selected: 1, options: [0.5, 0.6, 0.7, 0.8, 0.9, 1, 1.1, 1.25, 1.5, 1.75, 2] }
			};
			/*
			var controls = [
				'play-large', // The large play button in the center
				'restart', // Restart playback
				'rewind', // Rewind by the seek time (default 10 seconds)
				'play', // Play/pause playback
				'fast-forward', // Fast forward by the seek time (default 10 seconds)
				'progress', // The progress bar and scrubber for playback and buffering
				'current-time', // The current time of playback
				'duration', // The full duration of the media
				'mute', // Toggle mute
				'volume', // Volume control
				'captions', // Toggle captions
				'settings', // Settings menu
				'pip', // Picture-in-picture (currently Safari only)
				'airplay', // Airplay (currently Safari only)
				'download', // Show a download button with a link to either the current source or a custom URL you specify in your options
				'fullscreen', // Toggle fullscreen
			];
			*/

			var opts2 = {
				// count up instead of down
				invertTime: false,
				// show duration
				controls: ['play-large', 'play', 'progress', 'current-time', 'mute', 'volume', 'captions', 'settings', 'pip', 'airplay', 'fullscreen', 'duration']
			};

			me.player = new Plyr('#' + me.playerId, opts2);
			//me.player = new Plyr('#' + me.playerId);
			//me.player.play();
			//me.player.pause();
			
			me.player.on('ready', function(e) {
				if (me.plyrLog) {
					me.plyrLog("    Player initialized", 1);
					me.plyrLog("       ID: " + me.playerId, 1);
				}
				if (me.plyrOnLoaded) {
					if (me.plyrOnLoaded instanceof Function) {
						me.plyrOnLoaded(me.playerId, me.player);
					}
					else if (me.plyrOnLoaded instanceof Object)
					{
						if (me.plyrOnLoaded.fn) {
							if (me.plyrOnLoaded.scope) {
								Ext.Function.pass(me.plyrOnLoaded.fn, [ me.playerId, me.player ], me.plyrOnLoaded.scope)();
							}
							else {
								me.plyrOnLoaded.fn(me.playerId, me.player);
							}
						}
					}
				}
			});

			me.player.on('progress', function(e) {
				if (me.plyrLog) {
					me.plyrLog("    Player progress", 1);
					me.plyrLog("       ID: " + me.playerId, 1);
				}
				if (me.plyrOnProgress) {
					if (me.plyrOnProgress instanceof Function) {
						me.plyrOnProgress(me.playerId, me.player);
					}
					else if (me.plyrOnProgress instanceof Object)
					{
						if (me.plyrOnProgress.fn) {
							if (me.plyrOnProgress.scope) {
								Ext.Function.pass(me.plyrOnProgress.fn, [ me.playerId, me.player ], me.plyrOnProgress.scope)();
							}
							else {
								me.plyrOnProgress.fn(me.playerId, me.player);
							}
						}
					}
				}
			});
		}
	}
});
